File: services/tu-core/internal/domain/teacher/models/valueobjects/gender.go
package valueobjects

import "fmt"

// Gender adalah value object untuk jenis kelamin.
type Gender struct {
	code  string
	label string
}

var (
	GenderMale   = Gender{code: "MALE", label: "Laki-laki"}
	GenderFemale = Gender{code: "FEMALE", label: "Perempuan"}
)

// NewGender membuat Gender dari code.
func NewGender(code string) (Gender, error) {
	switch code {
	case "MALE":
		return GenderMale, nil
	case "FEMALE":
		return GenderFemale, nil
	default:
		return Gender{}, fmt.Errorf("gender tidak valid: %s", code)
	}
}

// Code mengembalikan code gender.
func (g Gender) Code() string { return g.code }

// Label mengembalikan label gender.
func (g Gender) Label() string { return g.label }

// Equals mengecek kesamaan.
func (g Gender) Equals(other Gender) bool { return g.code == other.code }

File: apps/siakad-tu/src/domain/student/value-objects/Gender.js
// apps/siakad-tu/src/domain/student/value-objects/Gender.js

/**
 * Value Object: Gender
 * Enum: MALE, FEMALE
 */
export class Gender {
  static MALE = new Gender('MALE', 'Laki-laki')
  static FEMALE = new Gender('FEMALE', 'Perempuan')

  static VALUES = [Gender.MALE, Gender.FEMALE]

  #code
  #label

  constructor(code, label) {
    this.#code = code
    this.#label = label
    Object.freeze(this)
  }

  get code() {
    return this.#code
  }
  get label() {
    return this.#label
  }

  equals(other) {
    if (!(other instanceof Gender)) return false
    return this.#code === other.#code
  }

  toString() {
    return this.#code
  }

  toJSON() {
    return this.#code
  }

  static fromJSON(code) {
    const found = Gender.VALUES.find((g) => g.#code === code)
    if (!found) {
      throw new Error(`Gender tidak valid: "${code}". Harus MALE atau FEMALE.`)
    }
    return found
  }
}
